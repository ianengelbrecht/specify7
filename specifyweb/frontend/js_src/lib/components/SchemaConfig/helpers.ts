import type { LocalizedString } from 'typesafe-i18n';

import { schemaText } from '../../localization/schema';
import type { IR, RA } from '../../utils/types';
import { addMissingFields } from '../DataModel/addMissingFields';
import type { SerializedResource } from '../DataModel/helperTypes';
import { parseJavaClassName } from '../DataModel/resource';
import type {
  JavaType,
  LiteralField,
  Relationship,
} from '../DataModel/specifyField';
import type { SpLocaleContainerItem, Tables } from '../DataModel/types';
import type { Aggregator, Formatter } from '../Forms/dataObjFormatters';
import type {
  ItemType,
  NewSpLocaleItemString,
  SpLocaleItemString,
} from './index';
import type { DataObjectFormatter } from './SetupHooks';

let newStringId = 1;
const defaultLanguage = 'en';
const defaultCountry = null;

export function findString(
  strings: RA<SpLocaleItemString> | undefined,
  language: string,
  country: string | null,
  itemType: 'containerDesc' | 'containerName' | 'itemDesc' | 'itemName',
  parentUrl: string
): NewSpLocaleItemString | SpLocaleItemString {
  /*
   * Start searching for matching string from the end. This would align
   * schema config behavior with the way back-end handles cases when there
   * are duplicate SpLocalteItemStr records for the same field and same language
   */
  const targetString = Array.from(strings ?? [])
    .reverse()
    .find(
      (object) =>
        object.language === language &&
        (object.country ?? '') === (country ?? '')
    );
  if (typeof targetString === 'object') return targetString;

  const defaultItem = strings?.find(
    (object) =>
      object.language === defaultLanguage && object.country === defaultCountry
  );
  newStringId += 1;

  return addMissingFields('SpLocaleItemStr', {
    id: -newStringId,
    text: defaultItem?.text ?? '',
    language,
    country,
    [itemType]: parentUrl,
  });
}

/** Throws away unneeded fields */
export const formatAggregators = (
  aggregators: RA<Aggregator | Formatter>
): IR<DataObjectFormatter> =>
  Object.fromEntries(
    aggregators.map(({ name = '', title = '', className = '' }) => [
      name,
      {
        title: title as LocalizedString,
        className,
      },
    ])
  );

/**
 * Filter down defined formatters
 */
export const filterFormatters = (
  formatters: IR<DataObjectFormatter>,
  tableName: keyof Tables
): IR<string> =>
  Object.fromEntries(
    Object.entries(formatters)
      .filter(
        ([_name, { className }]) =>
          parseJavaClassName(className).toLowerCase() === tableName
      )
      .map(([name, { title }]) => [name, title] as const)
  );

/**
 * Determine what kind of item SpLocalItem is based on what fields it has
 *
 * Assuming it can't be multiple types at once
 *
 */
export function getItemType(
  item: SerializedResource<SpLocaleContainerItem>
): ItemType {
  if (item.webLinkName !== null) return 'webLink';
  else if (item.pickListName !== null) return 'pickList';
  else if (item.format === null) {
    return 'none';
  } else {
    return 'formatted';
  }
}

const webLinkTypes = new Set<JavaType>(['text', 'java.lang.String']);

export function isFormatterAvailable(
  field: LiteralField | Relationship,
  formatter: ItemType
): boolean {
  if (formatter === 'none' || formatter === 'pickList') return true;
  else if (formatter === 'webLink')
    return !field.isRelationship && webLinkTypes.has(field.type);
  else if (formatter === 'formatted') return !field.isRelationship;
  else return false;
}

export const localizedRelationshipTypes: IR<string> = {
  'one-to-one': schemaText.oneToOne(),
  'one-to-many': schemaText.oneToMany(),
  'many-to-one': schemaText.manyToOne(),
  'many-to-many': schemaText.manyToMany(),
};

/**
 * Localize Java type name for presenting in the UI
 */
export function javaTypeToHuman(
  type: string | null,
  relatedModelName: string | undefined = ''
): string {
  if (type === null) return '';
  else if (type in localizedRelationshipTypes)
    return `${localizedRelationshipTypes[type]} (${relatedModelName})`;
  else if (type.startsWith('java')) return type.split('.').at(-1)!;
  else return type;
}
