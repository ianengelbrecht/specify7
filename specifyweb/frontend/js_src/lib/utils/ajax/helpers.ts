import type { IR, RA } from '../types';

// These HTTP methods do not require CSRF protection
export const csrfSafeMethod = new Set(['GET', 'HEAD', 'OPTIONS', 'TRACE']);

export const isExternalUrl = (url: string): boolean =>
  /*
   * Blob URLs may point to the same origin, but should be treated as external
   * by the navigator
   */
  url.startsWith('blob:') ||
  new URL(url, globalThis.location.origin).origin !==
    globalThis.location.origin;

// Make sure the given URL is from current origin and give back the relative path
export function toRelativeUrl(url: string): string | undefined {
  const parsed = new URL(url, globalThis.location.origin);
  return parsed.origin === globalThis.location.origin
    ? `${parsed.pathname}${parsed.search}${parsed.hash}`
    : undefined;
}

/**
 * Convert JSON object to FormData.
 * Some endpoints accept form data rather than stringified JSON.
 * Just wrap your JS object in a call to formData() before passing it as a
 * "body" to ajax()
 */
export function formData(
  data: IR<Blob | RA<number | string> | boolean | number | string>
): FormData {
  const formData = new FormData();
  Object.entries(data).forEach(([key, value]) =>
    formData.append(
      key,
      Array.isArray(value)
        ? JSON.stringify(value)
        : typeof value === 'number'
        ? value.toString()
        : typeof value === 'boolean'
        ? value.toString()
        : value
    )
  );
  return formData;
}
