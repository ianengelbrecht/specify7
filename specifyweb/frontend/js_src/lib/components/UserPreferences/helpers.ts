/**
 * Utilities for dealing with user preferences
 */

import { ajax } from '../../utils/ajax';
import { Http } from '../../utils/ajax/definitions';
import { ping } from '../../utils/ajax/ping';
import { cacheEvents, getCache, setCache } from '../../utils/cache';
import { f } from '../../utils/functools';
import { mergeParsers, parserFromType } from '../../utils/parser/definitions';
import { parseValue } from '../../utils/parser/parse';
import type { RA } from '../../utils/types';
import { filterArray, setDevelopmentGlobal } from '../../utils/types';
import { keysToLowerCase, replaceKey } from '../../utils/utils';
import { MILLISECONDS } from '../Atoms/Internationalization';
import { softFail } from '../Errors/Crash';
import {
  cachableUrl,
  contextUnlockedPromise,
  foreverFetch,
} from '../InitialContext';
import { formatUrl } from '../Router/queryString';
import type { Preferences } from './Definitions';
import { preferenceDefinitions } from './Definitions';
import { prefEvents } from './Hooks';

export function getPrefDefinition<
  CATEGORY extends keyof Preferences,
  SUBCATEGORY extends keyof Preferences[CATEGORY]['subCategories'],
  ITEM extends keyof Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items']
>(
  category: CATEGORY,
  subcategory: SUBCATEGORY,
  item: ITEM
): Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items'][ITEM] {
  const definition =
    // @ts-expect-error
    preferenceDefinitions[category].subCategories[subcategory].items[item];
  const defaultValue = defaultPreferences[category]?.[subcategory]?.[item];
  return defaultValue === undefined
    ? definition
    : replaceKey(definition, 'defaultValue', defaultValue);
}

/** Use usePref hook instead whenever possible as it comes with live updates */
export const getUserPref = <
  CATEGORY extends keyof Preferences,
  SUBCATEGORY extends CATEGORY extends keyof typeof preferenceDefinitions
    ? keyof Preferences[CATEGORY]['subCategories']
    : never,
  ITEM extends keyof Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items']
>(
  category: CATEGORY,
  subcategory: SUBCATEGORY,
  item: ITEM
): Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items'][ITEM]['defaultValue'] =>
  preferences[category]?.[subcategory]?.[item] ??
  defaultPreferences[category]?.[subcategory]?.[item] ??
  getPrefDefinition(category, subcategory, item).defaultValue;

let preferences: {
  readonly [CATEGORY in keyof Preferences]?: {
    readonly [SUBCATEGORY in keyof Preferences[CATEGORY]['subCategories']]?: {
      readonly [ITEM in keyof Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items']]?: Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items'][ITEM]['defaultValue'];
    };
  };
} = getCache('userPreferences', 'cached') ?? {};
export type UserPreferences = typeof preferences;
export const getRawUserPreferences = () => preferences;

export const setPrefsGenerator = (
  getPreferences: () => UserPreferences,
  triggerSync: boolean
) =>
  function setPref<
    CATEGORY extends keyof Preferences,
    SUBCATEGORY extends keyof Preferences[CATEGORY]['subCategories'],
    ITEM extends keyof Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items']
  >(
    category: CATEGORY,
    subcategory: SUBCATEGORY,
    item: ITEM,
    value: Preferences[CATEGORY]['subCategories'][SUBCATEGORY]['items'][ITEM]['defaultValue']
  ): void {
    const definition = getPrefDefinition(category, subcategory, item);

    let parsed;
    if ('type' in definition) {
      const baseParser = parserFromType(definition.type);
      const parser =
        typeof definition.parser === 'object'
          ? mergeParsers(baseParser, definition.parser)
          : baseParser;
      const parseResult = parseValue(
        parser,
        undefined,
        value?.toString(),
        parser.type !== 'text'
      );
      if (parseResult.isValid) parsed = parseResult.parsed;
      else {
        console.error(`Failed parsing pref value`, {
          category,
          subcategory,
          item,
          definition,
          parseResult,
        });
        parsed = definition.defaultValue;
      }
    } else if ('values' in definition) {
      if (definition.values.some((item) => item.value === value))
        parsed = value;
      else {
        console.error(`Failed parsing pref value`, {
          category,
          subcategory,
          item,
          value,
          definition,
        });
        parsed = definition.defaultValue;
      }
    } else parsed = value;

    const prefs = getPreferences() as any;
    if (
      parsed ===
      (prefs[category]?.[subcategory]?.[item] ?? definition.defaultValue)
    )
      return;

    prefs[category] ??= {};
    prefs[category][subcategory] ??= {};
    prefs[category][subcategory][item] = parsed;

    if (triggerSync) {
      /*
       * Unset default values
       * This reduces the size of the downloaded file, but mainly, it allows for
       * future Specify 7 versions to change the default value.
       */
      if (parsed === definition.defaultValue) {
        prefs[category][subcategory][item] = undefined;
        // Clean up empty objects
        if (
          filterArray(Object.values(prefs[category][subcategory])).length === 0
        )
          prefs[category][subcategory] = undefined;
        if (filterArray(Object.values(prefs[category])).length === 0)
          prefs[category] = undefined;
      }

      commitToCache();
      requestPreferencesSync();
    }
    prefEvents.trigger('update', {
      category,
      subcategory: subcategory as string,
      item: item as string,
      definition,
    });
  };
export const setPref = setPrefsGenerator(getRawUserPreferences, true);

// Sync with back-end at most every 5s
const syncTimeout = 5 * MILLISECONDS;
let syncTimeoutInstance: ReturnType<typeof setTimeout> | undefined = undefined;
let isSyncPending = false;
let isSyncing = false;

export async function awaitPrefsSynced(): Promise<void> {
  if (typeof syncTimeoutInstance === 'number') {
    globalThis.clearTimeout(syncTimeoutInstance);
    syncTimeoutInstance = undefined;
    return syncPreferences();
  }

  return isSyncing
    ? new Promise((resolve) => {
        const destructor = prefEvents.on('synchronized', () => {
          destructor();
          resolve();
        });
      })
    : Promise.resolve();
}

/** Update back-end with front-end changes in a throttled manner */
function requestPreferencesSync(): void {
  if (isSyncing) isSyncPending = true;
  else {
    if (typeof syncTimeoutInstance === 'number')
      globalThis.clearTimeout(syncTimeoutInstance);
    syncTimeoutInstance = globalThis.setTimeout(
      (): void => void syncPreferences().catch(softFail),
      syncTimeout
    );
  }
}

async function syncPreferences(): Promise<void> {
  await preferencesPromise;
  isSyncPending = false;
  return ping(
    `/context/user_resource/${userResource.id}/`,
    {
      method: 'PUT',
      body: keysToLowerCase({
        name: resourceName,
        mimeType: 'application/json',
        metaData: '',
        data: JSON.stringify(preferences),
      }),
    },
    {
      expectedResponseCodes: [Http.NO_CONTENT],
    }
  ).then(() => {
    // If there were additional changes while syncing
    if (isSyncPending) syncPreferences().catch(softFail);
    else {
      isSyncing = false;
      prefEvents.trigger('synchronized');
    }
  });
}

const resourceName = 'UserPreferences';
const defaultResourceName = 'DefaultUserPreferences';
const mimeType = 'application/json';

let userResource: ResourceWithData = undefined!;
let defaultPreferences: UserPreferences =
  getCache('userPreferences', 'defaultCached') ?? {};

type UserResource = {
  readonly id: number;
  readonly metadata: string | null;
  readonly name: string;
  readonly mimetype: string | null;
};
type ResourceWithData = UserResource & {
  readonly data: string;
};

/**
 * Fetch app resource that stores current user preferences
 *
 * If app resourcee data with user preferences does not exists does not exist,
 * check if SpAppResourceDir and SpAppResource exist and create them if needed,
 * then, create the app resource data itself
 */
export const preferencesPromise = contextUnlockedPromise.then(
  async (entrypoint) =>
    entrypoint === 'main'
      ? f
          .all({
            items: ajax<RA<UserResource>>(
              cachableUrl('/context/user_resource/'),
              {
                headers: { Accept: 'application/json' },
              }
            )
              .then(
                ({ data }) =>
                  data.find(
                    ({ name, mimetype }) =>
                      name === resourceName && mimetype === 'application/json'
                  )?.id
              )
              .then(async (appResourceId) =>
                (typeof appResourceId === 'number'
                  ? ajax<ResourceWithData>(
                      cachableUrl(`/context/user_resource/${appResourceId}/`),
                      {
                        headers: { Accept: 'application/json' },
                      }
                    )
                  : ajax<ResourceWithData>(
                      '/context/user_resource/',
                      {
                        headers: { Accept: 'application/json' },
                        method: 'POST',
                        body: keysToLowerCase({
                          name: resourceName,
                          mimeType,
                          metaData: '',
                          data: '{}',
                        }),
                      },
                      { expectedResponseCodes: [Http.CREATED] }
                    )
                ).then(({ data }) => data)
              ),
            defaultItems: ajax(
              formatUrl('/context/app.resource', {
                name: defaultResourceName,
                quiet: '',
              }),
              {
                headers: { Accept: 'text/plain' },
              },
              {
                expectedResponseCodes: [Http.NO_CONTENT, Http.OK],
                strict: false,
              }
            )
              .then(({ data, status }) =>
                status === Http.OK ? JSON.parse(data) : {}
              )
              .catch((error) => {
                softFail(error);
                return {};
              }),
          })
          .then(({ items, defaultItems }) => {
            defaultPreferences = defaultItems;
            initializePreferences(items);
            return items;
          })
      : foreverFetch<ResourceWithData>()
);

function initializePreferences(resource: ResourceWithData): ResourceWithData {
  userResource = resource;
  preferences = JSON.parse(userResource.data ?? '{}');
  setDevelopmentGlobal('_preferences', preferences);
  prefEvents.trigger('update', undefined);
  commitToCache();
  setCache('userPreferences', 'defaultCached', defaultPreferences);

  registerChangeListener();
  return userResource;
}

const commitToCache = (): void =>
  // Need to create a shallow copy of the resource since it can get mutated
  void setCache('userPreferences', 'cached', { ...preferences });

/** Listen for changes to preferences in another tab */
const registerChangeListener = (): void =>
  void cacheEvents.on('change', ({ category, key }) => {
    if (category !== 'userPreferences') return;
    if (key === 'cached')
      preferences = getCache('userPreferences', 'cached') ?? preferences;
    else if (key === 'defaultCached')
      defaultPreferences =
        getCache('userPreferences', 'defaultCached') ?? defaultPreferences;
    setDevelopmentGlobal('_preferences', preferences);
    prefEvents.trigger('update', undefined);
  });
