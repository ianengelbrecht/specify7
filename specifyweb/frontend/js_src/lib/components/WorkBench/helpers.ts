import { stringify } from 'csv-stringify/browser/esm';

import { downloadFile } from '../Molecules/FilePicker';
import { getUserPref } from '../UserPreferences/helpers';
import type { Dataset } from '../WbPlanView/Wrapped';

export const downloadDataSet = async ({
  name,
  rows,
  columns,
}: Dataset): Promise<void> =>
  new Promise((resolve, reject) =>
    stringify(
      [columns, ...rows],
      {
        delimiter: getUserPref('workBench', 'editor', 'exportFileDelimiter'),
      },
      (error, output) => {
        if (error === undefined)
          resolve(
            downloadFile(name.endsWith('.csv') ? name : `${name}.tsv`, output)
          );
        else reject(error);
      }
    )
  );
