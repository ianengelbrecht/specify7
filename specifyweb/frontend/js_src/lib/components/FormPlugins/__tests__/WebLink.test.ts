import { requireContext } from '../../../tests/helpers';
import { parseWebLink, webLinks } from '../WebLink';

requireContext();

test('weblinks are fetched and parsed correctly', async () => {
  const links = await webLinks;
  const parsed = Object.fromEntries(
    Object.entries(links).map(([name, definition]) => [
      name,
      parseWebLink(definition),
    ])
  );
  expect(parsed).toMatchSnapshot();
});
