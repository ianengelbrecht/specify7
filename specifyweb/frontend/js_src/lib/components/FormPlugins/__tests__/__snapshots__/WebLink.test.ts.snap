// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weblinks are fetched and parsed correctly 1`] = `
{
  "AccessionNumber": {
    "args": [
      "accessionNumber",
    ],
    "isExternal": false,
    "template": "http://specify6-test.nhm.ku.edu/echo.php?text1=<%= accessionNumber %>",
    "title": "",
  },
  "CTScan": {
    "args": [
      "_this",
    ],
    "isExternal": false,
    "template": "https://www.morphosource.org/biological_specimens/<%= _this %>",
    "title": "",
  },
  "FishBase": {
    "args": [
      "genus",
      "species",
    ],
    "isExternal": false,
    "template": "http://www.fishbase.org/Summary/<%= genus %>-<%= species %>.html",
    "title": "Displays a FishBase page for the Genus and Species",
  },
  "GUID": {
    "args": [
      "_this",
    ],
    "isExternal": false,
    "template": "urn:uuid:<%= _this %>",
    "title": "",
  },
  "Genbank": {
    "args": [
      "_this",
    ],
    "isExternal": false,
    "template": "http://www.ncbi.nlm.nih.gov/nuccore/<%= _this %>",
    "title": "",
  },
  "MailTo": {
    "args": [
      "email",
    ],
    "isExternal": false,
    "template": "mailto:<%= email %>",
    "title": "Launches an email message window for sending an email.",
  },
  "Specify7": {
    "args": [
      "catalogNumber",
    ],
    "isExternal": false,
    "template": "https://ichthyology.specify.ku.edu/specify/bycatalog/KUI/<%= catalogNumber %>/",
    "title": "",
  },
  "WEBLINK": {
    "args": [
      "_this",
    ],
    "isExternal": false,
    "template": "http://specify6-test.nhm.ku.edu/echo.php?text1=<%= _this %>",
    "title": "",
  },
}
`;
